File: bestia-zoneserver/src/main/java/net/bestia/zoneserver/actor/map/PlayerMoveRequestActor.java
package net.bestia.zoneserver.actor.map;

import java.util.Objects;

import org.springframework.beans.factory.annotation.Autowired;
import org.springframework.context.annotation.Scope;
import org.springframework.stereotype.Component;

import akka.actor.AbstractActor;
import net.bestia.entity.EntityService;
import net.entity.component.MoveComponent;
import net.bestia.messages.entity.EntityMoveRequestMessage;
import net.bestia.zoneserver.actor.zone.ClientMessageActor.RedirectMessage;

/**
 * Incoming player requests to move a bestia must be send towards the actor of
 * the entity which will handle the movement. Message musst be wrapped in an
 * component envelope in order to get delivered.
 * 
 * @author Thomas Felix
 *
 */
@Component
@Scope("prototype")
public class PlayerMoveRequestActor extends AbstractActor {

	private final EntityService entityService;

	@Autowired
	public PlayerMoveRequestActor(EntityService entityService) {

		this.entityService = Objects.requireNonNull(entityService);
	}

	@Override
	public Receive createReceive() {
		return receiveBuilder()
				.match(EntityMoveRequestMessage.class, this::handleMoveRequest)
				.build();
	}

	@Override
	public void preStart() throws Exception {
		// Register for chat commands.
		final RedirectMessage redirMsg = RedirectMessage.get(EntityMoveRequestMessage.class);
		getContext().parent().tell(redirMsg, getSelf());
	}

	private void handleMoveRequest(EntityMoveRequestMessage msg) {
		
		final MoveComponent mc = entityService.getComponentOrCreate(msg.getEntityId(), MoveComponent.class);
		mc.setPath(msg.getPath());
		entityService.updateComponent(mc);

	}
}

File: bestia-zoneserver/src/main/kotlin/net/bestia/zoneserver/actor/map/PlayerMoveRequestActor.kt
package net.bestia.zoneserver.actor.map

import net.bestia.zoneserver.entity.EntityService
import net.bestia.zoneserver.entity.component.MoveComponent
import net.bestia.messages.entity.EntityMoveMessage
import net.bestia.zoneserver.actor.routing.BaseClientMessageRouteActor
import org.springframework.context.annotation.Scope
import org.springframework.stereotype.Component

/**
 * Incoming player requests to move a bestia must be send towards the actor of
 * the entity which will handle the movement. Message must be wrapped in an
 * component envelope in order to get delivered.
 *
 * @author Thomas Felix
 */
@Component
@Scope("prototype")
class PlayerMoveRequestActor(
        private val entityService: EntityService
) : BaseClientMessageRouteActor() {

  override fun createReceive(builder: BuilderFacade) {
    builder.match(EntityMoveMessage::class.java, this::handleMoveRequest)
  }

  private fun handleMoveRequest(msg: EntityMoveMessage) {
    val moveComponent = MoveComponent.fromCollection(msg.entityId, msg.path)
    entityService.updateComponent(moveComponent)
  }
}